src/txrequest.cpp

```diff
@@ -0,0 +1,685 @@
+    size_t m_candidate_best = 0;
+    //! Number of REQUESTED entries for this txhash.
+    size_t m_requested = 0;
+    //! The priority of the CANDIDATE_BEST entry if one exists, or 0 otherwise.
+    uint64_t m_priority_candidate_best = 0;
+    //! The lowest priority of all CANDIDATE_READY entries (or max() if none exist).
+    uint64_t m_priority_best_candidate_ready = std::numeric_limits<uint64_t>::max();
+    //! All peers we have an entry for this txhash for.
+    std::vector<uint64_t> m_peers;
+};
+
+/** (Re)compute the PeerInfo map from the index. Only used for sanity checking. */
+std::unordered_map<uint64_t, PeerInfo> RecomputePeerInfo(const Index& index)
+{
+    std::unordered_map<uint64_t, PeerInfo> ret;
+    for (const Entry& entry : index) {
+        PeerInfo& info = ret[entry.m_peer];
+        ++info.m_total;
+        info.m_requested += (entry.GetState() == State::REQUESTED);
+        info.m_completed += (entry.GetState() == State::COMPLETED);
+    }
+    return ret;
+}
+
+/** Compute the TxHashInfo map. Only used for sanity checking. */
+std::map<uint256, TxHashInfo> ComputeTxHashInfo(const Index& index, const PriorityComputer& computer)
+{
+    std::map<uint256, TxHashInfo> ret;
+    for (const Entry& entry : index) {
+        TxHashInfo& info = ret[entry.m_txhash];
+        // Classify how many Entrys of each state we have for this txhash.
+        info.m_candidate_delayed += (entry.GetState() == State::CANDIDATE_DELAYED);
+        info.m_candidate_ready += (entry.GetState() == State::CANDIDATE_READY);
+        info.m_candidate_best += (entry.GetState() == State::CANDIDATE_BEST);
+        info.m_requested += (entry.GetState() == State::REQUESTED);
+        // And track the priority of the best CANDIDATE_READY/CANDIDATE_BEST entries.
+        if (entry.GetState() == State::CANDIDATE_BEST) {
+            info.m_priority_candidate_best = computer(entry);
+        }
+        if (entry.GetState() == State::CANDIDATE_READY) {
+            info.m_priority_best_candidate_ready = std::min(info.m_priority_best_candidate_ready, computer(entry));
+        }
+        // Also keep track of which peers this txhash has a Entry for (so we can detect duplicates).
+        info.m_peers.push_back(entry.m_peer);
+        // Track preferred/first.
+    }
+    return ret;
+}
+
+const uint256 UINT256_ZERO;
+
+}  // namespace
+
+/** Actual implementation for TxRequestTracker's data structure. */
+class TxRequestTracker::Impl {
+    //! The current sequence number. Increases for every announcement. This is used to sort txhashes returned by
+    //! GetRequestable in announcement order.
+    uint64_t m_sequence{0};
+
+    //! This tracker's priority computer.
+    const PriorityComputer m_computer;
+
+    //! This tracker's main data structure.
+    Index m_index;
+
+    //! Map with this tracker's per-peer statistics.
+    std::unordered_map<uint64_t, PeerInfo> m_peerinfo;
+
+public:
```

Comment:
It may be matter of personal taste.

In general I think it still makes sense to have non-exposed classes with private/public fields/members. It's not there to prevent external code from messing with the internals, obviously, but it does still kind of define a layer between what is part of the class's representation and what is its (even just internally) exposed interface.

In this case with the pimpl pattern... maybe that's silly, as the glue layer to forward calls on `TxRequestTracker` to `TxRequestTracker::Impl` is super thin, and it's already obvious what the exposed part is.

Happy to change it if there are strong opinions.

Reply:
No strong opinion. Feel free to mark this resolved.